vm/src/stdlib/socket.rs

```diff
@@ -234,10 +322,42 @@ impl PySocket {
         vm: &VirtualMachine,
     ) -> PyResult<()> {
         let flags = flags.unwrap_or(0);
-        let sock = self.sock();
-        bytes
-            .with_ref(|buf| py_io::write_all(buf, |b| sock.send_with_flags(b, flags)))
-            .map_err(|err| convert_sock_error(vm, err))
+
+        let timeout = self.timeout.load();
+        let timeout = if timeout > 0.0 {
+            Some(Duration::from_secs_f64(timeout))
+        } else {
+            None
+        };
+
+        let mut deadline: Option<Instant> = None;
+
+        let buf_len = bytes.len();
+        let mut buf_offset = 0;
+        // now we have like 3 layers of interrupt loop :)
+        while buf_offset < buf_len {
+            let interval = timeout.map(|dur| match deadline {
+                Some(d) => d
+                    .checked_duration_since(Instant::now())
+                    // past the deadline already
+                    .ok_or_else(|| timeout_error(vm)),
+                None => {
+                    let dl = Instant::now() + dur;
+                    deadline = Some(dl);
+                    Ok(dur)
```

Comment:
I am not understanding this part very well. `deadline` looks like to be None at the beginning and only assigned as Some here. If then, it looks like to decided out of loop. Am I missing somethinig?

Reply:
I think CPython does it like this to avoid the cost of the monotonic clock when there's no deadline. I think it would simplify the code a lot just to always do it; or should I just add comments about why we do it like this?